WIP: Initial implementation in progress

Leengit opened the pull request.

Commit:
ENH: Using ITK/Modules/Core/GPUCommon sources.

Commit:
ENH: Using Elastix sources instead.

Commit:
DOC: Adding questions and comments.  Formatting.  Adding documentation skeleton.
BUG: Compiling better.

Commit:
ENH: Adding type conversions for PyTorch.

Commit:
ENH: Removing OpenCL specifics.
ENH: Moving some methods from itkPyTorchDataManager to itkPyTorchImageDataManager because latter is aware of PixelType.
BUG: Making it compile.

aylward commented:
You should use MONAI instead of PyTorch.

S

---
Sent via phone.

Leengit commented:
aylward> You should use MONAI instead of PyTorch.

One purpose of this module is to supply `PyTorchImage` as a C++ template available for those using ITK in C++.  I am implementing the `PyTorchImage` template using the libtorch C++ interface.

Commit:
DOC: Adding more technical description of purpose and functionality.

Leengit commented:
It is intended that this module will integrate with MONAI.  If there are MONAI C++ bindings that I should be aware of, please let me know.  Or, if the approach of a creating a `PyTorchImage< TPixelType, VNumberOfDimensions >` template that publicly inherits from  the `Image<TPixelType, VNumberOfDimensions >` template is going to create incompatibilities with MONAI, please let me know.

aylward reviewed (commented):
Thanks for the clarification - I do think that perhaps MONAI should still be carefully considered, but perhaps you already are, and I am missing it in your code - you've done quite a bit of work already!

In general, nice implementation - easy to follow.   Enabling on-GPU pytorch image processing with ITK is outstanding.

My uncertainty regarding your code and MONAI will hopefully be cleared up when I see the jupyter notebook and/or better understand the primary target use case: is the goal implementing new transforms for pytorch using ITK or making use of pytorch networks in ITK python scripts?   Both seem feasible, but perhaps there are different considerations based on which is the goal...    

In particular, the target use case will help determine how image-to-physical-space information should be handled.  In MONAI, image-to-physical-space transforms (origin, spacing, and directions) are handled as metadata associated with an image.  From my quick review, I don't see that the standard used by MONAI is being adopted here, but perhaps I am wrong - the impact of that decision perhaps depends on the target use case - again, maybe I am just missing it.   Admittedly, I think MONAI's approach faces some challenges, so perhaps there is an opportunity to collaborate here.    MONAI is still in beta and welcomes majors changes when necessary.   Timing for working with them might be very good.   If you want to learn more, @hjmjohnson might have some insights for you.

Congrats on the progress already made and congrats for engaging folks for feedback!   Sorry for my overly brief response this morning.

Best regards,
Stephen

Commit:
DOC: Changing "Dirty" to "Stale" in variable names.
ENH: Infering dimensions within PixelType via DimensionHelper template.
BUG: Removing "override" declaration where superclass method is not virtual.
ENH: Creating test/itkPyTorchImageTest.cxx.
ENH: Builds if cmake is given -DCMAKE_PREFIX_PATH:PATH="/absolute/path/to/libtorch-cxx11-abi-shared-with-deps-1.5.1+cu101/libtorch".

dzenanz commented on src/itkPyTorchDataManager.cxx:
```diff
@@ -0,0 +1,275 @@
+  this->UpdateGPUBuffer();
+  this->UpdateCPUBuffer();
+
+  m_IsGPUBufferDirty = m_IsCPUBufferDirty = false;
```
I think it is better to split this into 2 explicit assignments:
```C++
m_IsGPUBufferDirty = false;
m_IsCPUBufferDirty = false;
```

Leengit replied:
Will do.

dzenanz commented on src/itkPyTorchDataManager.cxx:
```diff
@@ -0,0 +1,275 @@
+
+    m_CPUBuffer = data->m_CPUBuffer;
+    //    m_Platform  = data->m_Platform;
+    //    m_Context   = data->m_Context;
```
is this commented-out on purpose?

Leengit replied:
That's the obsolete OpenCL code waiting to be replaced by libtorch code.

dzenanz commented on include/itkPyTorchImage.h:
```diff
@@ -34,10 +34,10 @@ namespace itk
  */
 template< typename TPixel, unsigned int VImageDimension = 2 >
-class ITK_TEMPLATE_EXPORT PyTorchImage : public Image< TPixel, VImageDimension >
+class PyTorchImage : public Image< TPixel, VImageDimension >
```
We might need `ITK_TEMPLATE_EXPORT` specification. Its absence might be a problem in Elastix. @thewtex?

thewtex replied:
:+1: Yes, this macro is required

Leengit replied:
Will do.  I could use PYTORCH_EXPORT for PyTorchDataManager, which is not a template.  Except that PYTORCH_EXPORT is not found by the compiler.  Should I be defining it somewhere?

thewtex replied:
`PyTorch_EXPORT` can be used for non-templated classes, e.g. *TorchDataManager* (this is defined in the build tree in *PyTorchExport.h*, but for templates, always use `ITK_TEMPLATE_EXPORT`, which is defined in the ITK header *itkMacro.h*.

Leengit replied:
In the elastix module, I see the variable name and file name prefixed with ITK/itk and the module name is not in all upper case: `ITKOpenCL_Export` defined within `itkOpenCLExport.h`.  What should I do for the present case?

dzenanz replied:
You don't need to include anything, see https://github.com/InsightSoftwareConsortium/ITKMontage/blob/46c409f5b9f2a58a5cd76b77f5c90dfc2229a7c0/include/itkTileMontage.h#L45. `ITK_TEMPLATE_EXPORT` is defined in `itkMacro.h` which is probably transitively included by all the ITK headers.

dzenanz replied:
For the `Montage` module, the DLL export specifications are in `MontageExport.h` file. But those should be used only for non-templated classes and functions.

dzenanz commented on include/itkPyTorchImage.hxx:
```diff
@@ -15,45 +15,70 @@
-template <typename TPixel, unsigned int VImageDimension>
-PyTorchImage<TPixel, VImageDimension>::PyTorchImage()
+//
+// Constructor
```
This comment is not very helpful.

Leengit replied:
I'll disappear it.

thewtex commented:
@aylward thanks for sharing the awesome thoughts! Good integration with MONAI is certainly a priority, and the best way to do that will be informed and evolve according to suggestions from experts like yourself and @hjmjohnson.

I created this issue to track and collect ideas related to Jupyter notebooks and MONAI: https://github.com/InsightSoftwareConsortium/ITKPyTorch/issues/2

thewtex reviewed (commented):
@Leengit looking good.

While *ITKPyTorch* and *PyTorch* are good names for the module, I think the class names should have a *Torch* instead of *PyTorch* prefix. This will result in shorter names more consistent with the API in PyTorch itself, i.e. *torch* in the Python filter names.

dzenanz commented on include/itkPyTorchImage.hxx:
```diff
@@ -62,155 +87,259 @@ PyTorchImage<TPixel, VImageDimension>::Initialize()
+  {
+    // Attempt to cast data to an PyTorchImageDataManagerType
+    using PyTorchImageDataManagerType = PyTorchImageDataManager< PyTorchImage >;
+    const PyTorchImageDataManagerType * ptr;
```
maybe this should be initialized to `nullptr`

Leengit replied:
Will do.

dzenanz commented on include/itkPyTorchImageDataManager.hxx:
```diff
@@ -102,40 +87,50 @@ PyTorchImageDataManager<ImageType>::MakeCPUBufferUpToDate()
-    ModifiedTimeType gpu_time = this->GetMTime();
-    TimeStamp        cpu_time_stamp = m_Image->GetTimeStamp();
-    ModifiedTimeType cpu_time = m_Image->GetMTime();
+    unsigned long gpu_time       = this->GetMTime();
```
Why this change?

Leengit replied:
To be restored.

dzenanz commented on include/itkPyTorchImage.h:
```diff
@@ -134,106 +134,127 @@ class PyTorchImage : public Image< TPixel, VImageDimension >
   virtual void Initialize() override;
 
-  virtual void FillBuffer( const TPixel &value ) override;
+  virtual void FillBuffer( const TPixel &value ) /* `override` fails because not virtual in itkImage!!! */;
```
Then it probably should not be `virtual` here either?

Leengit replied:
I agree that they should be consistent.  I would appreciate knowing why they are not virtual in Image<...>.

aylward commented:
I do think this is a great direction to head!  Sorry if my suggestions are
off mark, but my limited understanding of this work suggests that perhaps
two specific considerations regarding MONAI:
1) You and others created some nice python helper functions for ITK (e.g.,
ITKExtras.py) that help with using ITK from numpy, accessing spacing
information, etc.   I wonder if something similar should be done in this
module for MONAI.
2) MONAI is using python dictionaries for managing spacing etc.   Seems
like those dictionaries could be created and populated by this toolkit as
well.  Yes, MONAI doesn't provide a custom C++ library for those
dictionaries, but does it need to in order to access them from C++?

Again, perhaps I am way off base, and I will drop it for now.  However, if
this helps with the proper handling of medical images by pytorch/monai,
then perhaps these additional investigations are worth it.

On Mon, Jul 20, 2020 at 10:05 AM Matt McCormick <notifications@github.com>
wrote:

> @aylward <https://github.com/aylward> thanks for sharing the awesome
> thoughts! Good integration with MONAI is certainly a priority, and the best
> way to do that will be informed and evolve according to suggestions from
> experts like yourself and @hjmjohnson <https://github.com/hjmjohnson> 's
> suggestions.
>
> I created this issue to track and collect ideas related to Jupyter
> notebooks and MONAI: #2
> <https://github.com/InsightSoftwareConsortium/ITKPyTorch/issues/2>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/InsightSoftwareConsortium/ITKPyTorch/pull/1#issuecomment-661060851>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACEJL4GEN7CJVVLCMI76QDR4RFKBANCNFSM4O436SZQ>
> .
>


-- 
Stephen R. Aylward, Ph.D.
Senior Director of Strategic Initiatives
---
Kitware: *Advancing the frontiers of understanding*
*by developing innovative open-source software platforms*
*and integrating them into research, processes, and products.*

thewtex commented:
@aylward definitely! Tracked in https://github.com/InsightSoftwareConsortium/ITKPyTorch/issues/3

Commit:
ENH: Supporting complex pixel types.
ENH: Computing PyTorchDimension at compile time.
ENH: Implementing GPUTensor memory management.
ENH: Using ITK_TEMPLATE_EXPORT.
BUG: Removing `virtual` from methods that are not virtual in the superclass.

Leengit commented:
> @Leengit looking good.
> 
> While _ITKPyTorch_ and _PyTorch_ are good names for the module, I think the class names should have a _Torch_ instead of _PyTorch_ prefix. This will result in shorter names more consistent with the API in PyTorch itself, i.e. _torch_ in the Python filter names.

Done.

Commit:
ENH: Changing file, class, and variable names, but not the module name, from using PyTorch to using Torch.

Commit:
ENH: Using cmake FetchContent for Torch (aka libtorch).

dzenanz commented on include/itkPyTorchImage.hxx:
```diff
@@ -57,11 +54,13 @@ void
 PyTorchImage< TPixel, VImageDimension >
 ::AllocateGPU()
 {
+  // Much of the same work is done in Initialize.  Where should it be done?!!!
```
It could be done in a protected/private method, which both `Initialize` and `Allocate` call.

dzenanz reviewed (commented):
Looks good so far

Commit:
ENH: Creating and using PyTorch_EXPORT macro
